### PR TITLE
Modular FNO

### PR DESCRIPTION
**Pull request by MarsuPila**
# Modulus Pull Request

## Description
This MR re-organises the FNO implementation in a modular way and makes encoders for the respective dimension transparent to outside modulus. That way, customisation of FNO is significantly simplified for the user.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/models/fno/__init__.py [hunk 1]
```diff
@@ -12,4 +12,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from .fno import FNO
+from .fno import FNO, FNO1DEncoder, FNO2DEncoder, FNO3DEncoder, FNO4DEncoder, MetaData
```
Comment (NickGeneva) [inline modulus/models/fno/__init__.py:15]: Is MetaData required in the init here?

File: modulus/models/fno/fno.py [hunk 2]
```diff
@@ -73,19 +73,25 @@ def __init__(
...
+        self.pad = padding[:1]
+        self.ipad = [-pad if pad > 0 else None for pad in self.pad]
+        self.padding_type = padding_type
 
-        # Initial lift network
+        # build lift
+        self.build_lift_network()
+        self.build_fno(num_fno_modes)
+
+    def build_lift_network(self) -> None:
```
Comment (ktangsali) [inline modulus/models/fno/fno.py:94]: Since these are not private, let's add docstrings to all the `build_lift_network`, `build_fno` etc. methods.

**Review by ktangsali — APPROVED** [review]
Looks good to me. Left a minor comment about the docstrings. Just to make sure we are getting the same results, let's also verify that the `darcy_fno` example predictions look good.

File: modulus/models/fno/fno.py (hunk 2 above)
Comment (MarsuPila) [reply modulus/models/fno/fno.py:97]: done!

**Comment by MarsuPila** [comment]
just started the `darcy_fno` example, hasn't finished yet, but after 90 epochs maximum relative error is below 3%

**Comment by MarsuPila** [comment]
> just started the `darcy_fno` example, hasn't finished yet, but after 90 epochs maximum relative error is below 3%

newplot
validation

**Review by NickGeneva — APPROVED** [review]

**Pull request merged**